kernel/obj_core.c

```diff
@@ -0,0 +1,357 @@
+			      size_t stats_len)
+{
+	__ASSERT((obj_core != NULL) && (stats != NULL), "NULL parameter");
+
+	k_spinlock_key_t  key = k_spin_lock(&lock);
+
+	__ASSERT(obj_core->type != NULL, "Object core not initialized");
+
+	if (obj_core->type->stats_desc == NULL) {
+
+		/* Object type not configured for statistics. */
+
+		k_spin_unlock(&lock, key);
+		return -ENOTSUP;
+	}
+
+	if (obj_core->type->stats_desc->raw_size != stats_len) {
+
+		/* Buffer size mismatch */
+
+		k_spin_unlock(&lock, key);
+		return -EINVAL;
+	}
+
+	obj_core->stats = stats;
+	k_spin_unlock(&lock, key);
+
+	return 0;
+}
+
+int k_obj_core_stats_deregister(struct k_obj_core *obj_core)
+{
+	__ASSERT((obj_core != NULL), "NULL parameter");
+
+	k_spinlock_key_t  key = k_spin_lock(&lock);
+
+	__ASSERT(obj_core->type != NULL, "Object core not initialized");
+
+	if (obj_core->type->stats_desc == NULL) {
+
+		/* Object type not configured  for statistics. */
+
+		k_spin_unlock(&lock, key);
+		return -ENOTSUP;
+	}
+
+	obj_core->stats = NULL;
+	k_spin_unlock(&lock, key);
+
+	return 0;
+}
+
+int k_obj_core_stats_raw(struct k_obj_core *obj_core, void *stats,
+			 size_t stats_len)
+{
+	int  rv;
+	struct k_obj_core_stats_desc *desc;
+
+	__ASSERT((obj_core != NULL) && (stats != NULL), "NULL parameter");
+
+	k_spinlock_key_t  key = k_spin_lock(&lock);
+
+	__ASSERT(obj_core->type != NULL, "Object core not initialized");
+
+	desc = obj_core->type->stats_desc;
+	if ((desc == NULL) || (desc->raw == NULL)) {
+
+		/* The object type is not configured for this operation */
+
+		k_spin_unlock(&lock, key);
+		return -ENOTSUP;
+	}
+
+	if ((desc->raw_size != stats_len) || (obj_core->stats == NULL)) {
+
+		/*
+		 * Either the size of the stats buffer is wrong or
+		 * the kernel object was not registered for statistics.
+		 */
+
+		k_spin_unlock(&lock, key);
+		return -EINVAL;
+	}
+
+	rv = desc->raw(obj_core, stats);
+	k_spin_unlock(&lock, key);
+
+	return rv;
+}
+
+int k_obj_core_stats_query(struct k_obj_core *obj_core, void *stats,
+			   size_t stats_len)
+{
+	int  rv;
+	struct k_obj_core_stats_desc *desc;
+
+	__ASSERT((obj_core != NULL) && (stats != NULL), "NULL parameter");
+
+	k_spinlock_key_t  key = k_spin_lock(&lock);
+
+	__ASSERT(obj_core->type != NULL, "Object core not initialized");
+
+	desc = obj_core->type->stats_desc;
+	if ((desc == NULL) || (desc->query == NULL)) {
+
+		/* The object type is not configured for this operation */
+
+		k_spin_unlock(&lock, key);
+		return -ENOTSUP;
+	}
+
+	if ((desc->query_size != stats_len) || (obj_core->stats == NULL)) {
+
+		/*
+		 * Either the size of the stats buffer is wrong or
+		 * the kernel object was not registered for statistics.
+		 */
+
+		k_spin_unlock(&lock, key);
+		return -EINVAL;
+	}
+
+	rv = desc->query(obj_core, stats);
+	k_spin_unlock(&lock, key);
+
+	return rv;
+}
+
+int k_obj_core_stats_reset(struct k_obj_core *obj_core)
+{
+	int  rv;
+	struct k_obj_core_stats_desc *desc;
+
+	__ASSERT((obj_core != NULL), "NULL parameter");
```

Comment:
Style: all this checking for NULL seems wasteful to me.  That's (1) a clear and obvious software bug in the app or subsystem and (2) really obvious in a debugger when it happens at runtime.  I don't think the assertion is providing much value, and promiscuous assertions tend over the long term to **discourage** people from using asserts  (because the app doesn't fit when it's enabled!).   Zephyr CONFIG_ASSERT has always been comparatively lightweight, this doesn't look lightweight to me.

Reply:
Good point. Will adjust.